Implement noopWindowsResourceAllocator

adelina-t opened the pull request:
On Windows, the podAdmitHandler returned by the GetAllocateResourcesPodAdmitHandler() func
and registered by the Kubelet is nil.

We should check for nil before registering any podAdmitHandlers.

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change

/kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #88913

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

benmoss commented:
@liggitt this is more fallout from #87759

liggitt commented:
Is there a test that demonstrates the bug this fixes?

adelina-t commented:
/test pull-kubernetes-e2e-aks-engine-azure-windows

adelina-t commented:
@liggitt ^

adelina-t commented:
/cc @michmike 
/cc @pjh

michmike reviewed (approved):
/lgtm
/approve

adelina-t commented:
/test pull-kubernetes-e2e-kind

dashpole commented:
/lgtm
/priority critical-urgent
looks urgent to me, since it causes failure to start on windows.

liggitt commented:
/assign @derekwaynecarr 
as approver of https://github.com/kubernetes/kubernetes/pull/87759

pjh commented:
/lgtm

Thanks Adelina!

mattjmcnaughton reviewed (approved):
/lgtm

Long term, another option would be for `AddPodAdmitHandler` to gracefully handle being passed `nil`? Definitely not blocking but curious if you thought about going that route?

liggitt commented:
/retest

mattjmcnaughton commented:
/retest

adelina-t commented:
/retest

adelina-t commented:
/retest

adelina-t commented:
/retest

liggitt commented:
given the test windows cluster has failed to start on multiple consecutive runs, is there further investigation that needs to be done before retest?

adelina-t commented:
Windows cluster failed over the weekend two times because of staging error (?) , nothing to do with the cluster:

```
W0308 21:07:52.765] 2020/03/08 21:07:52 process.go:153: Running: /go/src/k8s.io/release/push-build.sh --nomock --verbose --noupdatelatest --bucket=kubernetes-release-pull --ci --gcs-suffix=/pull-kubernetes-e2e-aks-engine-azure-windows --allow-dup
I0308 21:07:52.965] push-build.sh: BEGIN main on 6af94723-617f-11ea-b188-56b122f30985 Sun Mar  8 21:07:52 UTC 2020
I0308 21:07:52.965] 
W0308 21:07:53.066] tar (child): /go/src/k8s.io/kubernetes/_output/release-tars/kubernetes.tar.gz: Cannot open: No such file or directory
W0308 21:07:53.066] tar (child): Error is not recoverable: exiting now
W0308 21:07:53.067] tar: Child returned status 2
W0308 21:07:53.067] tar: Error is not recoverable: exiting now
I0308 21:07:53.167] 
I0308 21:07:53.168] 
I0308 21:07:53.168] Signal ERR caught!
```

Today it failed because a timeout in installing features on one of the Windows nodes:

```
Add-WindowsFeature : The server could not update the provided feature files in the time allowed.
```

None of these failures are related to this code.

adelina-t commented:
With latest Windows presubmit run, Kubelet starts on Windows nodes, so this PR is doing it's job. The test failures are related to the fact that there is a missing agnhost image: e2eteam/agnhost:2.12 . @claudiubelu  just pushed it so I'll triggered a retest.

/retest

adelina-t commented:
/retest

adelina-t commented:
/retest

chewong commented:
@adelina-t I think https://github.com/kubernetes/test-infra/pull/16537 broke the presubmit job. Let me modify the job configuration accordingly.

chewong commented:
/retest

chewong commented:
/retest

adelina-t commented:
There are only 2 tests failing in the latest presubmit aks-engine job run, both of them unrelated to this patch. We should really merge this so we can have the Kubelet up and running ( and the e2e tests ) so we can determine if the two failing tests are flakes or if they are really an issue.

liggitt commented:
GetAllocateResourcesPodAdmitHandler() is not documented to be able to return nil, and all other implementations return a non-nil interface. It seems less fragile to have the windows version return a no-op implementation as well

michmike commented:
@liggitt, @adelina-t  will make the change you requested. Once done, we would appreciate a quick approval since this fix impacts a lot of windows tests that are failing.

adelina-t committed:
Implement noopWindowsResourceAllocator

On Windows, the podAdmitHandler returned by the GetAllocateResourcesPodAdmitHandler() func
and registered by the Kubelet is nil.

We implement a noopWindowsResourceAllocator that would admit any pod for Windows in order
to be consistent with the original implementation.

adelina-t commented:
/test pull-kubernetes-e2e-aks-engine-azure-windows

liggitt commented on pkg/kubelet/cm/container_manager_windows.go:
```diff
@@ -178,7 +186,7 @@ func (cm *containerManagerImpl) ShouldResetExtendedResourceCapacity() bool {
 
 func (cm *containerManagerImpl) GetAllocateResourcesPodAdmitHandler() lifecycle.PodAdmitHandler {
-	return nil
+        return &noopWindowsResourceAllocator{}
```
`hack/update-gofmt.sh` might not be happy with the whitespace change

liggitt commented:
/approve
/lgtm

change looks good to me. check if `hack/update-gofmt.sh` fixes the whitespace and push an update to fix that if so

liggitt commented:
also, per https://github.com/kubernetes/kubernetes/pull/88917#issuecomment-597243384, drop `Fixes: #88913` from the commit description (you can move it into the PR description)

adelina-t commented:
/test pull-kubernetes-e2e-aks-engine-azure-windows

liggitt commented:
/lgtm
/approve

adelina-t commented:
/test pull-kubernetes-verify

neolit123 commented:
/lgtm
based on the latest force push diff:
https://github.com/kubernetes/kubernetes/compare/e2caa2b93cab0ce21df6e18f5775323641d8ac35..a9f834d17dc990bd3ba282b51d784a567ecb86d6

/restest

neolit123 commented:
/retest

adelina-t commented:
/test pull-kubernetes-integration

adelina-t commented:
/test pull-kubernetes-integration

adelina-t commented:
/test pull-kubernetes-integration

adelina-t commented:
/test pull-kubernetes-e2e-kind

adelina-t commented:
/test pull-kubernetes-e2e-kind
/test pull-kubernetes-e2e-gce
/test pull-kubernetes-kubemark-e2e-gce-big
